CMakeLists.txt

```diff
@@ -0,0 +1,285 @@
+
+  add_custom_target(
+      plist ALL
+      DEPENDS $ENV{CPP_SRC_HOME}/mk/make_info_plist.sh 
+      DEPENDS $ENV{CPP_SRC_HOME}/gradle.properties
+      COMMAND $ENV{CPP_SRC_HOME}/mk/make_info_plist.sh "${ML_APP_NAME}" false > Info.plist
+  )
+  install(FILES ${CMAKE_BINARY_DIR}/Info.plist DESTINATION ${CMAKE_INSTALL_PREFIX} COMPONENT Distributed)
+  install(FILES $ENV{CPP_SRC_HOME}/lib/core/ml-en.dict DESTINATION ${CMAKE_INSTALL_PREFIX}/Resources COMPONENT Distributed)
+
+endif()
+
+if (CMAKE_SYSTEM_NAME STREQUAL "Linux")
+  message(STATUS "Linux detected")
+  set(DYNAMIC_LIB_DIR lib)
+  set(IMPORT_LIB_DIR lib)
+  # linux flags
+
+  if(${CMAKE_CROSSCOMPILING})
+    message(STATUS "CROSS_TARGET_PLATFORM=${CROSS_TARGET_PLATFORM}")
+    message(STATUS "CMAKE_CXX_STANDARD ${CMAKE_CXX_STANDARD}")
+    set(SYSROOT /usr/local/sysroot-${CROSS_TARGET_PLATFORM})
+    set(CROSS_FLAGS "--sysroot=${SYSROOT}")
+  endif()
+
+  # Target architecture specific flags
+  message(STATUS "CMAKE_SYSTEM_PROCESSOR ${CMAKE_SYSTEM_PROCESSOR}")
+  if (${CMAKE_SYSTEM_PROCESSOR} MATCHES "^aarch" OR ${CMAKE_SYSTEM_PROCESSOR} MATCHES "^arm")
+    message(STATUS "aarch64 detected for target")
+    set(ARCHCFLAGS "-march=armv8-a+crc+crypto")
+    add_compile_definitions(RAPIDJSON_NEON)
+  else()
+    message(STATUS "x86_64 detected for target")
+    set (ARCHCFLAGS "-msse4.2 -mfpmath=sse")
+    add_compile_definitions(RAPIDJSON_SSE42)
+  endif()
+  
+  set(CMAKE_C_FLAGS "-g ${CROSS_FLAGS} ${OPTCFLAGS} ${ARCHCFLAGS} -fstack-protector -fno-math-errno -Wall -Wcast-align -Wconversion -Wextra -Winit-self -Wno-psabi -Wparentheses -Wpointer-arith -Wswitch-enum")
+  set(CMAKE_CXX_FLAGS "${CMAKE_C_FLAGS} -Wno-ctor-dtor-privacy -Wno-deprecated-declarations -Wold-style-cast -fvisibility-inlines-hidden")
+
+  set(Boost_COMPILER "gcc${CMAKE_CXX_COMPILER_VERSION_MAJOR}")
+  set(EXE_DIR bin)
+  set(PLATFORM_NAME "Linux")
+  add_compile_definitions(Linux)
+
+  set(LIBXML2_HINTS /usr/local/gcc103)
+
+  # Hardcode paths to avoid any confusion
+  set(TORCH_INC "/usr/local/gcc103/include/pytorch")
+  set(TORCH_LIB "/usr/local/gcc103/lib/libtorch_cpu.so")
+  set(C10_LIB   "/usr/local/gcc103/lib/libc10.so")
+  set(ZLIB_LIBRARIES "-lz")
+  set(LIBXML2_LIBRARIES "-lxml2")
+
+  include_directories(SYSTEM ${SYSROOT}/usr/include)
+  include_directories(SYSTEM ${SYSROOT}/usr/local/include)
+  include_directories(SYSTEM ${SYSROOT}/usr/local/gcc103/include/libxml2)
+
+  install(FILES $ENV{CPP_SRC_HOME}/lib/core/ml-en.dict DESTINATION ${CMAKE_INSTALL_PREFIX}/resources COMPONENT Distributed)
+endif()
+
+if (CMAKE_SYSTEM_NAME STREQUAL "Windows")
+
+  message(STATUS "VCVER $ENV{VCVER}")
+  STRING(REGEX MATCH "^([0-9]+)\\.([0-9]+)\\.([0-9]+)" VCVER_REGEX_MATCH  $ENV{VCVER})
+  set(VCVER_MAJOR ${CMAKE_MATCH_1})
+  set(VCVER_MINOR ${CMAKE_MATCH_2})
+
+  string(CONCAT BOOST_VCVER "vc" ${VCVER_MAJOR} ${VCVER_MINOR} )
+  string(SUBSTRING ${BOOST_VCVER} 0 5 BOOST_VCVER)
```

Comment:
Why is the substring needed here?

If major is 14 and minor is 2 then surely the concat alone produces the right result.

If minor is not just 2 then I am wondering if there's some subtle future problem lurking here.

Reply:
The minor version is 20 in this case. However the Boost naming scheme only appears to use the first digit of it - see [version-numbers-for-visual-studio-2017-boost-and-cmake ](https://stackoverflow.com/questions/42730478/version-numbers-for-visual-studio-2017-boost-and-cmake#:~:text=And%20here%20it%20seems%20that,talk%20about%20boost's%20compiler%20naming.)